Clean up and improve Docker setup

MonkeyDo opened the pull request:
### Problem
 Docker files were old and borked, and the setup required a lot of faffing about.


### Solution
Clean up Dockerfile and docker-compose, set up scripts for database init and development.

mayhem commented on README.md:
```diff
@@ -120,32 +74,56 @@ To clone the repository and point the local HEAD to the latest commit in the
+If all went well, you will only need to run `./develop.sh` in the command line from the `bookbrainz-site` folder.
+Press `ctrl+c` to stop the server. The dependencies will continue running in the background.
+
+Wait between 30 and 60 seconds for the server to compile, and point your browser to `localhost:9099`.
```
Hmmm, not very portable. A shitty net connection or slow machines blows this out of the water. Sadly, this is very difficult to describe what happens, because the container logs don't really go silent, which is how I would describe when its good to go....

leftmostcat commented:
What's happened to the production `Dockerfile`? Why have the SQL files been copied into this repository? Is the plan to archive bookbrainz-sql and use only this repo? I would prefer to have this done as a submodule, as it makes sense that someone might want an instance of the BB database without wanting an instance of the site.

shivam-tripathi commented on MANUAL_INSTALL.md:
```diff
@@ -0,0 +1,73 @@
+
+## Elasticsearch
+
+To install Elasticsearch, follow [this helpful guide](https://www.digitalocean.com/community/tutorials/how-to-install-and-configure-elasticsearch-on-ubuntu-16-04) for Linux-based systems or the [official instructions](
```
Would be good to mention the version here for ES.

shivam-tripathi commented on src/server/helpers/auth.js:
```diff
@@ -50,46 +54,55 @@ function _updateCachedMBName(bbUserModel, mbUserJSON) {
-	});
+		passport.deserializeUser((user, done) => {
+			done(null, user);
+		});
```
Extra tabs?

MonkeyDo replied:
Checked in my code editor, looks fine to me.
These lines are in a new try/catch block, so they're indented one more.

shivam-tripathi commented on sql/scripts/dump.sh:
```diff
@@ -0,0 +1,29 @@
+#!/usr/bin/env bash
+
+# Swtich directory
```
Typo

shivam-tripathi commented on scripts/create-dumps.sh:
```diff
@@ -0,0 +1,43 @@
+
+# NOTE NOTE NOTE NOTE NOTE NOTE NOTE NOTE NOTE NOTE NOTE NOTE NOTE NOTE NOTE NOTE 
+# This is a saved script that is running in production. We don't know where
+this might be kept in repo, so we'e checking a version in here for safe keeping.
```
Noted, but it's missing a hash

MonkeyDo commented:
@leftmostcat 
> What's happened to the production `Dockerfile`?

I modified the existing Dockerfile that was out of date. It was the only Dockerfile.
I've removed the unused prod docker-compose & consul files; we have them in the history if we need them, but I will most likely be recreating them when it's time to move to the Hetzner architecture. They stale too, and were a possible source of confusion.

> Why have the SQL files been copied into this repository? Is the plan to archive bookbrainz-sql and use only this repo? I would prefer to have this done as a submodule, as it makes sense that someone might want an instance of the BB database without wanting an instance of the site.

That is indeed the plan. Both for local and production usage, it's a PITA and confusing to have the SQL files separate, for the benefit of a rare scenario that I suppose hasn't come up yet.
The goal is also to bring BB closer to other MeB projects' modus operandi to simplify deployment and troubleshooting by third parties that aren't BB developer (say zas for example).
To me that has more value than saving someone the trouble of cloning the site repo rather than the sql repo.

MonkeyDo commented on scripts/create-dumps.sh:
```diff
@@ -0,0 +1,43 @@
+
+DUMP_FILE=bookbrainz-dump-`date -I`.sql
+
+export PGPASSWORD="postgres"
```
Yes, yes it should. This has been copied from the current dump system and will be reviewed as part of moving to the Hetzner architecture

MonkeyDo commented on README.md:
```diff
@@ -120,32 +74,57 @@ To clone the repository and point the local HEAD to the latest commit in the
-## Building and running
-A number of subcommands exist to manage the installation and run the server.
-These are described here; any commands not listed should not be called directly:
+- change the bookbrainz-site service command to `npm run debug-watch-server`
```
Actually, this is intended for compiling with webpack inside docker.
I am rewriting that section to be clearer.

MonkeyDo commented on config/config.json.example:
```diff
@@ -4,30 +4,30 @@
+			"host": "postgres",
 			"database": "bookbrainz",
 			"user": "bookbrainz",
 			"password": "bookbrainz"
```
I hadn't realized they weren't, thanks for catching that!

MonkeyDo commented on scripts/create-db.sh:
```diff
@@ -0,0 +1,4 @@
+#!/bin/bash
+
+createuser -h postgres -U postgres bookbrainz
```
@mayhem Do you remember user bookbrainz being used anywhere?
It's not the case in the DB dump script or in any script in this repo

mayhem replied:
This is how other projects do it -- each DB has one or more users to access the DB. The DB dumps are not generating this username yet, since the production DB has not been moved to that user. This is one more step of making things more consistent -- the final step will come when the production DB runs under the bookbrainz user at Hetzner.

MonkeyDo replied:
Roger, thanks.

MonkeyDo commented on README.md:
```diff
@@ -34,76 +34,30 @@ Auto-generated developer documentation can be found at our corresponding
+When that is installed, clone the repository and follow the instructions below step by step.
 
-This will set the password to your PostgreSQL, which will be used for the database config.
+If you wish, you can instead [install the database and search dependencies on your machine](./MANUAL_INSTALL.md), and/or [run the NodeJS server locally](./NODEJS_SETUP.md) while using dockerized dependencies.
```
At this point, the only advantage I can think of is running locally for debugging purposes.

I've been running with file watching and hot-reloading inside of the docker container.
I lack the knowledge to do a useful resource comparison, though.

As time passes I'll get a better feel as to what the advantages are.

MonkeyDo commented on scripts/database-init-docker.sh:
```diff
@@ -0,0 +1,7 @@
+docker-compose build &&
+docker-compose pull &&
+docker-compose up -d postgres &&
+docker-compose exec bookbrainz-site scripts/create-db.sh &&
```
Thanks, I've corrected that and will push soon.

LordSputnik commented:
@MonkeyDo, @leftmostcat - I'm not sure I like deprecating bookbrainz-sql either. If somebody wants to run a python script to do something cool with the database, they shouldn't need to check out bookbrainz-site. The same reasoning is why we shouldn't have the database auto-generated from a nodejs ORM (as seems to be possible with Sequalize or TypeORM). It's good to be modular - and I think that the complexity of some other MeB repositories (for example, musicbrainz-server) puts people off contributing. I think the simplicity of our repositories and code is one of the reasons that we've had such a high level of contribution to BookBrainz in GCI in past years (that plus it's a younger project with more "low-hanging fruit").

mayhem commented:
> I think the simplicity of our repositories and code is one of the reasons that we've had such a high level of contribution to BookBrainz in GCI in past years

I'm sorry, but this is simply not the case. When Monkey expressed some frustration over getting the docker stuff to work, I offered to help for the day. As I was getting ready to dive in I had to spend a bunch of time finding things. Nothing was ever in a place where it should be. Nothing was organized like other projects -- BB had to be inexplicably different from the rest of MeB projects. Unless Monkey was there to guide me what to look at and what to check out, I would've walked away in frustration.

I think repositories need to have a good reason to exist, much the same way that we have good reasons for communication channels to exist. If you have too many, not all get used and the critical mass of a project is lost. Repos are the same way -- we need a good reason to split things into a multiple repos.

A vague "someone might want to check this out separately from the other stuff in a few years" loses out to newbies to our projects are having a hard time finding things. If you have a compelling reason, grounded in technical background and not personal preference, let me hear it!

paramsingh commented:
Just my two cents, I have a bunch of MetaBrainz projects that I help out with and building them didn't take me a long time because we follow the same scheme consistently. I have personally been avoiding building bookbrainz too especially because it isn't consistent with almost all our other projects and it would take me a long time to internalize the new stuff. Considering the fact that a lot of our projects are very similar data driven projects, if someone contributes in one of them, we should ideally be able to ask them for some kind of help in the other projects too. We're able to do this for CB, AB and LB but due to BB being completely different, it's hard to do for BB. BB contributors have to internalize a new workflow to contribute to other projects and vice versa. I would prefer consistency in all our projects.

leftmostcat commented:
Using a git submodule, which is easily documentable and a standard git workflow, allows us to retain the separability of the database and the site without, I hope, significant impact to discoverability. It also prevents us from losing repo history. I agree that our repos aren't simple (nor is our code), but it's also simpler to combine than it is to separate, and there's no way I know of to carry over repository history, either with separation or combination.

mayhem commented:
> which is easily documentable and a standard git workflow

Why isn't it already documented? It is a confusing workflow, especially for a beginner.

> retain the separability of the database

Why is that a goal? For a nascent project, this hampers getting into the project. 

Again, I would like to discover actual benefits that we get from keeping them separate. Any sort of argument of "in the future, this and that" are not valid -- that is premature optimisation. 

So far I've not heard any arguments that extend beyond premature optimistion or personal preference. Consistency with other projects alone tops all of the arguments I've heard so far.

MonkeyDo commented:
I am going to merge this now, after merging the bookbrainz-sql repo while conserving its history.
I am going to deprecate the bookbrainz-sql and bookbrainz-docker repos

MonkeyDo commented:
SQL files have been imported while conserving commit history
